### PR TITLE
kv support cleos

### PR DESCRIPTION
**Pull request by kimjh2005**
## Change Description

Cleos needs to be updated to use this new get_kv_table_rows API similarly to how the current cleos get table command uses get_table_rows.

## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

Syntax)
> cleos get kv_table --help
Retrieve the contents of a database kv_table
Usage: bin/cleos get kv_table [OPTIONS] account table index_name

Positionals:
  account TEXT REQUIRED       The account who owns the table
  table TEXT REQUIRED         The name of the kv_table as specified by the contract abi
  index_name TEXT REQUIRED    The name of the kv_table index as specified by the contract abi

Options:
  -h,--help                   Print this help message and exit
  -l,--limit UINT             The maximum number of rows to return
  -k,--key TEXT               Deprecated
  -L,--lower TEXT             JSON representation of lower bound value of key, defaults to first
  -U,--upper TEXT             JSON representation of upper bound value of key, defaults to last
  --encode-type TEXT          The encoding type of index_value, lower bound, upper bound 'bytes' for arbitrary hexdecimal bytes 'dec' for decimal encoding string of (uint64t , uint128_t , float64, float128) 'hex' for hexdecimal encoding string
  -b,--binary                 Return the value as BINARY rather than using abi to interpret as JSON
  -r,--reverse                Iterate in reverse order
  --show-payer                Show RAM payer
ex)
 cleos get kv_table  --encode-type name -L boba -U bobj  kvtable kvtable primarykey
 cleos get kv_table  --encode-type name -k boba   kvtable kvtable primarykey
 cleos get kv_table  --encode-type hex -k 1   kvtable kvtable foo
 cleos get kv_table  --encode-type dec -k 1   kvtable kvtable foo
 cleos get kv_table  --encode-type dec -L 2 -U 5   kvtable kvtable foo
 cleos get kv_table -r -b  --encode-type dec -L 2 -U 5   kvtable kvtable foo

### CODE REVIEW

File: plugins/chain_plugin/chain_plugin.cpp [hunk 1]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+   const string &upper_bound = *p.upper_bound;
+
    name database_id = chain::kvram_id;
    auto& gp = db.get_global_properties();
    auto& kv_config = gp.kv_configuration;
    const chain::kv_database_config& limits = kv_config.kvram;
 
-   auto&database = db.db();
+   auto& database = db.db();
    auto kv_context = chain::create_kv_chainbase_context(const_cast<chainbase::database&>(database), database_id, chain::name{0}, {}, limits);
```
Comment (linhuang-blockone) [inline plugins/chain_plugin/chain_plugin.cpp:2081]: Do we need to do for this RocksBd backing store?

File: tests/get_kv_table_nodeos_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   p.table = "kvtable"_n;
+   p.index_name = "primarykey"_n;
+   p.index_value = "boba";
+   p.encode_type = "name";
+   p.lower_bound = "";
+   p.upper_bound = "";
+   p.json = true;
+   p.reverse = false;
+   eosio::chain_apis::read_only::get_table_rows_result result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(1u, result.rows.size());
```
Comment (linhuang-blockone) [inline tests/get_kv_table_nodeos_tests.cpp:71]: Can you verify the row content too?

File: tests/get_kv_table_tests.cpp [hunk 3]
```diff
@@ -315,13 +315,16 @@ struct test_range_query_kv_context : public test_empty_kv_context {
       for( auto &prefix : prefixes ) {
          size_t prefix_size = prefix.size();
          vector<unsigned char> key;
-         key.resize(prefix_size + 4);
+         key.resize(prefix_size + 6);
```
Comment (linhuang-blockone) [inline tests/get_kv_table_tests.cpp:318]: What does 6 mean?

**Review by linhuang-blockone — CHANGES_REQUESTED** [review]

File: plugins/chain_plugin/chain_plugin.cpp [hunk 4]
```diff
@@ -2070,23 +2223,27 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
       auto cmp = itr->kv_it_key_compare(end_key.data(), end_key.size());
       if( reverse ) cmp *= -1;
 
+      fc::variant row_var;
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp:2226]: Notr used

File: plugins/chain_plugin/chain_plugin.cpp (hunk 1 above)
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp:2081]: This is change for  #367

File: tests/get_kv_table_nodeos_tests.cpp (hunk 2 above)
Comment (kimjh2005) [reply tests/get_kv_table_nodeos_tests.cpp:71]: There is a bug in abi_serializer that I need to fix in #367 work. I will add them in #367 work.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 5]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+   constexpr int dec_base = 10;
+
+   try {
+      // converts arbitrary hex strings to bytes ex) "FFFEFD" to {255, 254, 253}
+      if( encode_type == "bytes" ) {
+         if( bin.capacity() < index_value.size() ) {
+            bin.reserve(index_value.size());
+         }
+         string bytes_data = boost::algorithm::unhex(index_value);
+         auto bytes_datasize = bytes_data.size();
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp:1923]: changed

File: plugins/chain_plugin/chain_plugin.cpp [hunk 6]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+         for( char c : bytes_data ) {
+            bin.push_back(c);
+         }
+         return;
+      }
+
+      if( encode_type == "string" ) {
+         if( bin.capacity() < index_value.size() ) {
+            bin.reserve(index_value.size() + 2); // two zeros at the end
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp]: removed. No need to check size for convert_to_key

File: plugins/chain_plugin/chain_plugin.cpp [hunk 7]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+         }
+         convert_to_key(index_value, bin);
+         return;
+      }
+
+      if ( index_type == "name" ) {
+         name nm(index_value);
+         convert_to_key(nm.to_uint64_t(), bin);
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp:1938]: no check needed for convert_to_key

File: plugins/chain_plugin/chain_plugin.cpp [hunk 8]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+         }  else if( index_type == "int8" ) {
+            int8_t i8 = std::stoul(index_value, &pos, dec_base);
+            if( pos == 0 ) throw std::invalid_argument("invalid index_value " + index_value);
+            convert_to_key(i8, bin);
+         } else {
+            EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported index type/encode_type: ${t}/${e} ", ("t", index_type)("e", encode_type));
+         }
+      }
+   } catch( const std::invalid_argument& e) {
+      EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported index type/encode_type: ${t}/${e} {$v} ", ("t", index_type)("e", encode_type)("v", index_value));
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp]: changed

File: programs/cleos/main.cpp [hunk 9]
```diff
@@ -2872,6 +2872,46 @@ int main( int argc, char** argv ) {
...
+   // get kv_table
+   string index_name;
+   string index_value;
+   encode_type = "bytes";
+   auto getKvTable = get->add_subcommand("kv_table", localized("Retrieve the contents of a database kv_table"));
+   getKvTable->add_option( "account", code, localized("The account who owns the table") )->required();
+   getKvTable->add_option( "table", table, localized("The name of the kv_table as specified by the contract abi") )->required();
+   getKvTable->add_option( "index_name", index_name, localized("The name of the kv_table index as specified by the contract abi") )->required();
+   getKvTable->add_option( "-l,--limit", limit, localized("The maximum number of rows to return") );
+   getKvTable->add_option( "-k,--key", index_value, localized("Deprecated") );
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply programs/cleos/main.cpp]: removed copy and paste error

File: plugins/chain_plugin/chain_plugin.cpp (hunk 4 above)
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp]: removed

File: tests/get_kv_table_tests.cpp (hunk 3 above)
Comment (kimjh2005) [reply tests/get_kv_table_tests.cpp]: added  a comment

File: plugins/chain_plugin/chain_plugin.cpp [hunk 10]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+         convert_to_key(nm.to_uint64_t(), bin);
+         return;
+      }
+
+      size_t pos = 0;
+      if( encode_type == "hex" ) {
+         if( index_type == "sha256" || index_type == "i256" ) {
+            checksum256_type sha{index_value};
+            const char* sha_data = sha.data();
+            for( int i = 0; i < sha.data_size(); ++i ) {
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp]: used memcpy

File: plugins/chain_plugin/chain_plugin.cpp [hunk 11]
```diff
@@ -1911,33 +1912,172 @@ read_only::get_table_rows_result read_only::get_table_rows( const read_only::get
...
+      // converts arbitrary hex strings to bytes ex) "FFFEFD" to {255, 254, 253}
+      if( encode_type == "bytes" ) {
+         if( bin.capacity() < index_value.size() ) {
+            bin.reserve(index_value.size());
+         }
+         string bytes_data = boost::algorithm::unhex(index_value);
+         auto bytes_datasize = bytes_data.size();
+         for( char c : bytes_data ) {
+            bin.push_back(c);
+         }
```
[the remark this answers is not in the document]
Comment (kimjh2005) [reply plugins/chain_plugin/chain_plugin.cpp:1926]: used memcpy

**Review by linhuang-blockone — APPROVED** [review]

**Pull request merged**